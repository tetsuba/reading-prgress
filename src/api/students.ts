import axios from 'axios'
import {
    ApiResponseStudentTypes,
    ApiPayloadStudentTypes,
    ApiPayloadUpdateStudentTypes,
    BaseUrlType
} from './api-types'
import { getAuthHeaders, getBaseUrl } from './api-utils'

const URL_STUDENT: `${BaseUrlType}/student` = `${getBaseUrl()}/student`
const URL_REGISTER: `${typeof URL_STUDENT}/register` = `${URL_STUDENT}/register`
const URL_DELETE: `${typeof URL_STUDENT}/delete` = `${URL_STUDENT}/delete`
const URL_UPDATE: `${typeof URL_STUDENT}/update` = `${URL_STUDENT}/update`

export async function registerStudent(
    json: ApiPayloadStudentTypes
): Promise<ApiResponseStudentTypes> {
    return await axios.post(URL_REGISTER, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json',
            ...getAuthHeaders()
        }
    })
}

export async function deleteStudent(
    studentId: number
): Promise<ApiResponseStudentTypes> {
    return await axios.delete(`${URL_DELETE}?studentId=${studentId}`, {
        headers: {
            'Content-Type': 'application/json',
            ...getAuthHeaders()
        }
    })
}

export async function updateStudent(
    json: ApiPayloadUpdateStudentTypes
): Promise<ApiResponseStudentTypes> {
    return await axios.patch(URL_UPDATE, JSON.stringify(json), {
        headers: {
            'Content-Type': 'application/json',
            ...getAuthHeaders()
        }
    })
}
