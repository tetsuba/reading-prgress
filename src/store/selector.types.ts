import { StateCollectionTypes } from './store.types'
import { ApiTypes, BookTypes } from '../api/api-types'

type SelectorTypes = {
    iconColor: string
    numberOfBooks: number
    completed: boolean
}

export type SelectorBookTypes = Pick<BookTypes, 'id' | 'title'> &
    Pick<SelectorTypes, 'iconColor'>

export type SelectorCollectionTypes = Pick<StateCollectionTypes, 'title'> &
    Record<'books', SelectorBookTypes[]>

export type SelectorCollectionsTypes = Pick<
    StateCollectionTypes,
    'title' | 'id'
> &
    Pick<SelectorTypes, 'numberOfBooks' | 'completed'>

export type SelectorStudentLastProgressTypes =
    | undefined
    | Array<
          Pick<ApiTypes, 'collectionId' | 'bookId' | 'date'> &
              Pick<SelectorTypes, 'completed'>
      >
