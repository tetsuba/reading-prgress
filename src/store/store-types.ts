import {
    ApiBookHistoryTypes,
    ApiCollectionTypes,
    ApiStudentType
} from '../api/api-types'
import store from './store'

export type ActionTypes = {
    type: string
}

export type StateBookHistoryTypes = ApiBookHistoryTypes
export type StateBooksTypes = ApiCollectionTypes[]
export type StateStudentTypes = ApiStudentType

export interface StateBookTypes {
    bookId: number
    libId: string
    history: StateBookHistoryTypes[] | []
    story: string | string[]
    title: string
}

export interface StateUserTypes {
    firstName: string
    lastName: string
    email: string
    id: number
    token: string
}

export type BooksViewTypes = {
    collection: ApiCollectionTypes | null
}

export type StateViewGlobalTypes = {
    expired: boolean
}

export interface StateViewTypes {
    books: BooksViewTypes
    global: StateViewGlobalTypes
    studentId: number | null
}

export interface StateTypes {
    book: StateBookTypes
    books: StateBooksTypes
    students: StateStudentTypes[]
    user: StateUserTypes
    view: StateViewTypes
}

export type AppDispatch = typeof store.dispatch
