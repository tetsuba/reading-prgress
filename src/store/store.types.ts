import {
    ApiBookHistoryTypes,
    ApiCollectionTypes,
    ApiProgressType,
    ApiStudentType
} from '../api/api-types'
import store from './store'

export type ActionTypes = {
    type: string
}

export type StateBookHistoryTypes = ApiBookHistoryTypes
export type StateBooksTypes = ApiCollectionTypes[]
export type StateStudentTypes = ApiStudentType
export type StateCollectionTypes = ApiCollectionTypes
export type StateProgressTypes = ApiProgressType

export type StateBookTypes = {
    bookId: number
    libId: string
    history: StateBookHistoryTypes[] | []
    story: string | string[]
    title: string
}

export type StateUserTypes = {
    firstName: string
    lastName: string
    email: string
    id: number
    token: string
}

export type StateViewGlobalTypes = {
    expired: boolean
}

export type StateViewTypes = {
    global: StateViewGlobalTypes
    books: {
        showMessage: boolean
    }
}

export type StateCurrentTypes = {
    studentId: number | null
    collectionId: string | null
    bookId: number | null
}

export type StateTypes = {
    books: StateBooksTypes
    current: StateCurrentTypes
    students: StateStudentTypes[]
    user: StateUserTypes
    view: StateViewTypes
}

export type AppDispatch = typeof store.dispatch

export type ReadWordsTypes = {
    word: string
    index: number
}

export type StudentProgressReadWordsTypes = {
    oneWeekAgo: ReadWordsTypes[] | string[]
    oneMonthAgo: ReadWordsTypes[] | string[]
    history: ReadWordsTypes[] | string[]
}
