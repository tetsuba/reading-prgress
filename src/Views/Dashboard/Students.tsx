import { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import * as R from 'ramda'

// STORE
import { studentsSelector } from '../../store/students/studentsSelectors'
import { updateViewStudentId } from '../../store/view/viewSlice'

// TYPES
import { StateStudentTypes } from '../../store/store-types'

// COMPONENTS
import Loop from '../../Components/Loop/Loop'
import StudentRow from '../../Components/Row/StudentRow'
import ModalConfirmationDeleteStudent from './ModalConfirmationDeleteStudent'

export default function Students() {
    const dispatch = useDispatch()
    const students = useSelector(studentsSelector)
    const [deleteStudent, setDeleteStudent] =
        useState<StateStudentTypes | null>(null)

    return (
        <section>
            <Loop array={students}>
                <StudentRow
                    selectStudent={(student) => {
                        dispatch(updateViewStudentId(student.studentId))
                    }}
                    deleteStudent={(student) => {
                        setDeleteStudent(student)
                    }}
                />
            </Loop>
            <ModalConfirmationDeleteStudent
                student={deleteStudent}
                closeModal={() => setDeleteStudent(null)}
                show={R.isNotNil(deleteStudent)}
            />
        </section>
    )
}
