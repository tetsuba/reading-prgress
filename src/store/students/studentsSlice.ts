import { createSlice } from '@reduxjs/toolkit'
import { ActionTypes, StateStudentTypes } from '../store-types'

interface AddStudentActionTypes extends ActionTypes {
    payload: StateStudentTypes[]
}

export const initialState = [] as StateStudentTypes[] | []

export const studentsSlice = createSlice({
    name: 'students',
    initialState,
    reducers: {
        addStudents: (state, action: AddStudentActionTypes) => {
            return action.payload
        }
    }
})

export const { addStudents } = studentsSlice.actions

export default studentsSlice.reducer
