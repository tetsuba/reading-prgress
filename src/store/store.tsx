import { configureStore } from '@reduxjs/toolkit'
import { userSlice } from './user/userSlice'
import { bookSlice } from './book/bookSlice'
import { viewSlice } from './view/viewSlice'
import { booksSlice } from './books/booksSlice'
import { studentsSlice } from './students/studentsSlice'
import { studentSlice } from './student/studentSlice'

const store = configureStore({
    reducer: {
        book: bookSlice.reducer,
        books: booksSlice.reducer,
        student: studentSlice.reducer,
        students: studentsSlice.reducer,
        user: userSlice.reducer,
        view: viewSlice.reducer
    },
    devTools: true
})
export default store

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch
