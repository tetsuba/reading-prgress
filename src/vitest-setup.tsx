import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
