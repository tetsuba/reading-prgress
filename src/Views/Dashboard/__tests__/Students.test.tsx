import {
    render,
    screen,
    fireEvent,
    waitFor,
    waitForElementToBeRemoved
} from '@testing-library/react'
import { WrapperWith_Store_Query_Router } from '../../../vitest-setup'
import Students from '../Students'
import store from '../../../store/store'
import { addStudents } from '../../../store/students/studentsSlice'
import axios from 'axios'
import { Mocked } from 'vitest'

vi.mock('axios')
const mockedAxios = axios as Mocked<typeof axios>

const mockStudents = [
    {
        firstname: 'Alan',
        lastname: 'Ball',
        dob: '202-12-12',
        studentId: 1,
        progress: []
    },
    {
        firstname: 'Bob',
        lastname: 'Ball',
        dob: '202-12-12',
        studentId: 2,
        progress: []
    }
]

mockedAxios.delete.mockResolvedValue({
    data: [mockStudents[1]]
})

describe('Students', () => {
    store.dispatch(addStudents(mockStudents))

    test('should render students', () => {
        const { asFragment } = render(
            <WrapperWith_Store_Query_Router pathname={'/dashboard'}>
                <Students />
            </WrapperWith_Store_Query_Router>
        )
        expect(asFragment()).toMatchSnapshot()
        expect(screen.getAllByTestId('student-delete')).toHaveLength(2)
    })

    test('should delete a student', async () => {
        render(
            <WrapperWith_Store_Query_Router pathname={'/dashboard'}>
                <Students />
            </WrapperWith_Store_Query_Router>
        )
        expect(screen.getAllByTestId('student-delete')).toHaveLength(2)
        fireEvent.click(screen.getAllByTestId('student-delete')[0])
        await waitFor(() => expect(screen.getByTestId('modal-confirmation')))
        fireEvent.click(screen.getByTestId('delete-button'))
        await waitForElementToBeRemoved(() =>
            screen.getByTestId('modal-confirmation')
        )
        expect(screen.getAllByTestId('student-delete')).toHaveLength(1)
    })
    test('should select a student', () => {
        const name = `${mockStudents[0].firstname} ${mockStudents[0].lastname}`
        render(
            <WrapperWith_Store_Query_Router pathname={'/dashboard'}>
                <Students />
            </WrapperWith_Store_Query_Router>
        )
        fireEvent.click(screen.getAllByTestId('select-student')[0])
        expect(store.getState().view.studentId).toEqual(
            mockStudents[1].studentId
        )
    })
})
