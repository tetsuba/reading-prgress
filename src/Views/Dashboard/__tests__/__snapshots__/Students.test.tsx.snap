// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Students > should render students 1`] = `
<DocumentFragment>
  <section>
    <div
      class="flex justify-between px-4 py-5 sm:gap-4 sm:px-6 bg-gray-50"
    >
      <div
        class="flex items-center"
      >
        <span
          class="mr-6 text-gray-300"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.75 6a3.75 3.75 0 11-7.5 0 3.75 3.75 0 017.5 0zM4.501 20.118a7.5 7.5 0 0114.998 0A17.933 17.933 0 0112 21.75c-2.676 0-5.216-.584-7.499-1.632z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
        <span
          class="font-medium text-gray-800"
        >
          Alan Ball
        </span>
      </div>
      <div
        class="flex justify-end"
      >
        <button
          class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-gray bg-gray-200 hover:bg-gray-300 "
          data-testid="select-student"
        >
          Select
        </button>
        <button
          class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white ml-2"
          data-testid="student-delete"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="flex justify-between px-4 py-5 sm:gap-4 sm:px-6 bg-white"
    >
      <div
        class="flex items-center"
      >
        <span
          class="mr-6 text-gray-300"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.75 6a3.75 3.75 0 11-7.5 0 3.75 3.75 0 017.5 0zM4.501 20.118a7.5 7.5 0 0114.998 0A17.933 17.933 0 0112 21.75c-2.676 0-5.216-.584-7.499-1.632z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </span>
        <span
          class="font-medium text-gray-800"
        >
          Bob Ball
        </span>
      </div>
      <div
        class="flex justify-end"
      >
        <button
          class="flex items-center rounded-lg font-medium text-base text-center px-5 py-2.5 text-gray bg-gray-200 hover:bg-gray-300 "
          data-testid="select-student"
        >
          Select
        </button>
        <button
          class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white ml-2"
          data-testid="student-delete"
        >
          <svg
            class="h-6 w-6"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.74 9l-.346 9m-4.788 0L9.26 9m9.968-3.21c.342.052.682.107 1.022.166m-1.022-.165L18.16 19.673a2.25 2.25 0 01-2.244 2.077H8.084a2.25 2.25 0 01-2.244-2.077L4.772 5.79m14.456 0a48.108 48.108 0 00-3.478-.397m-12 .562c.34-.059.68-.114 1.022-.165m0 0a48.11 48.11 0 013.478-.397m7.5 0v-.916c0-1.18-.91-2.164-2.09-2.201a51.964 51.964 0 00-3.32 0c-1.18.037-2.09 1.022-2.09 2.201v.916m7.5 0a48.667 48.667 0 00-7.5 0"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </section>
</DocumentFragment>
`;
